man/dbdupd.1
.TH DBDUPD 1 "12 September 1995"
.FB
.SH NAME
dbdupd \- define a \fIFirstBase\fP database update dictionary
.SH SYNOPSIS
.B dbdupd
[\fB-d\fP \fIdatabase\fP] [\fB-i\fP \fIindex\fP]
.SH DESCRIPTION
.I Dbdupd
allows definition and modification of a 
.I FirstBase
database update dictionary. 
The created dictionary will later be used by
\fIdbugen\fP(1)
to generate updates within a \fIFirstBase\fP database.
.PP
A database update dictionary consists of 
a series of update descriptors. Each update descriptor 
contains a
.I field_name,
and an
.I update-string.
.PP
Database updates can be thought of as global changes to an entire section
of an existing database. The idea is to describe an action to be done
to a particular field (or fields) and then use 
\fIdbugen\fP(1)
to actually implement the update.
Fields not mentioned in the update dictionary are left intact.
.PP
NOTE: Another \fIFirstBase\fP tool, \fIdbmacro\fP(1), might
be more applicable than \fIdbdupd\fP and \fIdbugen\fP.
.PP
Normally, field names are used to build selection criteria.
But, if the CHOOSEFIELD environment variable is ON, fields can be selected
from a list of numbered entries. See \fIsetup\fP(5).
.PP
The type of action that
can be done on a field
depends on the type of the field as seen by the \fIFirstBase\fP system (i.e. the type
of the field as defined in the database dictionary).
Any field can be cleared (to spaces) by leaving the 
.I update-string
unspecified.
Also, alphanumeric fields can be 
.I replaced 
by the 
.I update-string.
.PP
But numeric fields allow full formula specification.
In this case the 
.I update-string 
becomes the formula for the update.
The new field value will be 
dictated by this formula and the contents of the fields referenced by the
formula.
This formula has
the same format as those for \fIFirstBase\fP formula (virtual) fields defined
via \fIdbdbas\fP(1).
.PP
Formulas 
use numbers to represent fields (eg. 16 for field 16)
and a prepended 'C' or 'c' for constants (eg. C4.5 for 4.5).
Addition 
.I (+), 
subtraction 
.I (-), 
multiplication 
.I (*), 
and division
.I (/)
are allowed as operators. 
.PP
.I Note:
Only dollar, float, numeric and positive numeric
fields can be used in formula
calculations. This means references to other formulas will not work.
Also, all formulas read from left to right, and there is no precedence
among the operators.
.PP
The precision of a formula is specified by appending a colon 
.I (:)
to the formula followed by a number.
Default precision is two.
.PP
All formula entry is parsed to determine whether the formula is valid
for the current database
dictionary. Interactive error messages help debug long
formulas.
.PP
For example, if a group of records (as defined by an index, of course),
represents all people who have paid a standard payment on a large bill,
then the new value of each balance can be globally
updated all at one time using 
\fIdbdupd\fP(1) and \fIdbugen\fP(1).
.PP
Using this case as an example, say the dollar balance due 
field is field number 5 and the payment is $116.92, then the 
.I update-string
would be \fB5 + C11692:0\fP.
.PP
Again, 
particular attention must be paid to the update string if the field to be
updated is numeric in nature. Specifically, if the precision of 
constants and other numbers are not set properly, the results will not
be as expected.
.PP
For example, to update a dollar field to contain $123.45, the actual update
string required is "C12345:0", since \fIFirstBase\fP 'knows' where the decimal goes for
all dollar fields. Dates are done as textual substitutions, "051660"
for example to get May 16, 1960.
.PP
A special meta-update string, \fB$DELETE\fP is used to delete
all records that are touched during running of the update generator.
In this case, the actual contents of the field are not modified.
Note: due to the nature of \fIdbugen\fP(1), the \fB$DELETE\fP meta-update
string can only be designated for alphanumeric fields.
.PP
Like all other \fIFirstBase\fP dictionary editors, \fIdbdupd\fP(1) has a set of commands
that allow screen oriented modification of the update dictionary.
The following is a summary of the commands accepted by 
\fIdbdupd\fP(1):
.TP 7
.B <CTL>-H
displays a summary of these commands.
.TP 7
.B ?[dfhiN]
displays help on the database, formulas,
user defined help files, index, or field N.
.TP 7
.B <RETURN>
(return, or enter) pages forward one page in the database
dictionary (same as 'f').
Wrapping occurs at end of file.
.TP 7
.B @
go into auto add mode.
.TP 7
.B #
allow changes to line number #.
.TP 7
.B -
standard END key \(em exits program and keeps all changes.
.TP 7
.B <CTL>-X
aborts the program. The effect is to exit without
overwriting the file, meaning all changes are lost.
.TP 7
.B b
pages backward one page in the database update dictionary. 
Wrapping occurs at beginning of file.
.TP 7
.B d
delete function. Allows deletion of a single dictionary entry.
All entries below deleted entry slide up one slot.
.TP 7
.B i
insert function. Inserts a standard dictionary entry before any given
field.
All entries below inserted entry slide down one slot.
.PP
A list of simple editor commands appears on the 24th line of the 
screen at all times, with help available by using the standard help 
signal (<CTL>-H).
Complete interactive error checking is done on the contents of
all update descriptors.
.PP
Upon successful completion of an update dictionary, the user is given a
chance to run the update generator. All arguments given
to \fIdbdupd\fP(1) are passed on to \fIdbugen\fP(1).
The processes are overlaid,
meaning no new shells are generated in using this option.
.SH FILES
.PD 0
.TP 10
dbase
default name for the database
.TP 10
index
default name for the index
.TP 10
*.idictu
dictionary for update generator.
.TP 10
*.cdb
\fIFirstBase\fP database.
.TP 10
*.map
record map of \fIFirstBase\fP database.
.PD
.SH SEE ALSO
dbedit(1), dbugen(1), dbmacro(1), firstbase(5), input(5), setup(5)
.PP
.I FirstBase User's Guide and Reference Manual
.br
.SH DIAGNOSTICS
\fIDbdupd\fP(1)
will allow editing and modifying of any update dictionary at any time.
These dictionary files should be write protected for complete protection.
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/dbugen/dbdupd.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: dbdupd.c,v 9.1 2001/02/16 19:48:39 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char sccsid[] = "@(#)dbdupd.c	8.2 05 Apr 1996 FB";
#endif

/* 
 *  dbdupd.c - screen editor for update dictionary (idictu) file
 */

#include <fb.h>
#include <fb_vars.h>
#include <dbd.h>

extern short int cdb_locklevel;
extern short int cdb_choosefield;
extern char *cdb_T_DELETE;
extern short int cdb_askgen;
extern char *cdb_UGEN;

#define HLP_DBDUPD	"dbdupd.hlp"

static int dot = 0;			/* current fb_field */
static int lastitem = 0;		/* last list item */

static fb_upd **itemlist = NULL;	/* for list of items */

static int col[8] = {1, 6, 18, 22, 30};	/* edit columns (screen) */

static fb_database *hp;

#define SHOWP 4				/* ending of shown screen columns */


static edit_idictu();
static fillscreen();
static escreen();
static void fdelete();
static void finsert();
static fadder();
static eitem();
static check();
static fput();
static newdictu();
static headers();

/* 
 *  dbdupd - main driver
 */
 
   main(argc, argv)
      int argc;
      char *argv[];
      
      {
         int st, i;
	 char idictu[FB_MAXNAME];

         cdb_locklevel = -1;
         fb_getargs(argc, argv, FB_ALLOCDB);
	 hp = cdb_db;
	 fb_initdbd(hp);
         if (fb_getd_dict(hp) == FB_ERROR)
            fb_xerror(FB_BAD_DICT, hp->ddict, NIL);            
         else
            for (i = 0; i < hp->nfields; i++)
               fb_nounders(cdb_keymap[i]);
	 fb_scrhdr(hp, "Define Updates");
         headers();
	 sprintf(idictu, "%su", hp->idict);	/* idictu file for updates */
         st = edit_idictu(idictu);
         if (st == FB_END){
            newdictu(idictu);
	    if (cdb_askgen)
	       fb_chainto("Generate Updates", cdb_UGEN, argv);
	    }
         fb_ender();
      }

/* 
 *  edit_idictu - edit an index.idictu or allow creation of one 
 */
 
   static edit_idictu(p)
      char *p;
      
      {
         int st, i;

	 itemlist =
	     (fb_upd **) fb_malloc((hp->nfields+1) * sizeof(fb_upd *));
	 			/* allocate enough ptrs for all fields */
	 if ((lastitem = read_idictu(p, itemlist)) == FB_ERROR)
	    lastitem = 0;
	 for (i = lastitem + 1; i < hp->nfields ; i++)
	    itemlist[i] = NULL;
         dot = 1;
         fillscreen();
         while ((st = escreen()) != FB_END)
            if (st == FB_ABORT)
               break;
            else if (st >= dot && st <= dot + 9 && st <= lastitem){
               fb_cx_push_env("6D", CX_KEY_SELECT, NIL);
               eitem(st, 0);
               fb_cx_pop_env();
               }
            else if (st >= 1 && st <= lastitem){
               dot = fb_onpage(st);
               fillscreen();
               }
            else
               fb_screrr("Invalid Input");
         return(st);
      }

/* 
 *  fillscreen - fill a screen with items
 */
 
   static fillscreen()
      {
         int j, i, p;
         
         fb_move(4, 1); fb_clrtobot();
         j = dot + 9;
         for (i = dot; i <= j && i <= lastitem; i++)
            for (p = 0; p <= SHOWP; ++p)
               fput(i, p);
	 fb_infoline();
         fb_cx_set_viewpage(dot);
      }
            
/* 
 *  escreen - set up a screen, accept commands for screen control 
 */
 
   static escreen()
      {
         int st;
	 char com[6];
         
         for (;;){
            fb_cx_push_env("E1", CX_DBD_SELECT, NIL);
            if (lastitem > 10)
               fb_cx_add_buttons("Ge");
            fb_cx_add_buttons("XH");
            st = fb_anychange(com);
            fb_cx_pop_env();
            switch (st) {
               case FB_ABORT:
	          if (fb_mustbe('y', "Really Quit? (y/n) ",
                        cdb_t_lines, 1) == FB_AOK){
                     fb_screrr("Update Dictionary File Unchanged");
		     return(st);
		     }
		  break;
               case ADDMODE:
	          if (lastitem < hp->nfields)
                     fadder();
		  else
		     continue;
                  break;
               case DELMODE:
	          if (lastitem > 0)
                     fdelete();
		  else
		     continue;
                  break;
               case FB_END:
                  return(st);
	       case HELPMODE:
	          fb_move(2, 1); fb_clrtobot();
		  fb_help(com, hp);
		  fb_move(4,1); fb_clrtobot();
		  headers();
		  break;
	       case INSERTMODE:
	          if (lastitem > 0 && lastitem < hp->nfields)
	             finsert();
		  else
		     continue;
		  break;
               case PAGEF:
                  dot = dot + 10;
                  if (dot > lastitem)
                     dot = 1;
                  break;
               case PAGEB:
                  dot = dot - 10;
                  if (dot < 1)
                     dot = fb_onpage(lastitem);
                  break;
	       case FB_QHELP: 
	          fb_move(2, 1); fb_clrtobot();
	          fb_fhelp(HLP_DBDUPD);
		  headers();
		  break;
	       case FB_AOK:
                  return(atoi(com));
               default:
                  return(st);
               }
            fillscreen();
            }
      }

/* 
 *  fdelete - fb_delete an update item 
 */
 
   static void fdelete()
      {
         int st, n;
         
         fb_move(cdb_t_lines, 1);
         fb_clrtoeol();
         fb_printw("Delete what number ? (#,-) ");
         st = fb_input(cdb_t_lines, 28, 3, 0, FB_INTEGER, (char *) &n,
            FB_ECHO, FB_OKEND, FB_CONFIRM);
         if (st == FB_END || st == FB_ABORT)
            return;
	 else if (n < 1 || n > lastitem)
	    return;
	 if (itemlist[n] != NULL)
	    fb_free((char *) itemlist[n]);
         for (; n < lastitem; n++)
            itemlist[n] = itemlist[n + 1];
         lastitem--;
      }

/* 
 *  finsert - insert a defaulted item before a given item 
 */
 
   static void finsert()            
      {
         int st, n, rn;

	 fb_fmessage("Insert before what number (#,-) ? ");
         st = fb_input(cdb_t_lines, 36, 3, 0, FB_INTEGER, (char *) &n,
            FB_ECHO, FB_OKEND, FB_CONFIRM);
         if (st == FB_END || st == FB_ABORT)
            return;
	 else if (n < 1 || n > lastitem)
	    return;
	 rn = n;
         for (n = lastitem; n >= rn; n--)		/* go backwards */
            itemlist[n+1] = itemlist[n];
	 itemlist[rn] = (fb_upd *) fb_malloc(sizeof(fb_upd));
	 itemlist[rn]->fp = NULL;		/* default to first fb_field */
	 itemlist[rn]->fv = NULL;
         lastitem++;
      }

/* 
 *  fadder - go into auto add mode 
 */
 
   static fadder()
      {
         int st;

         dot = fb_onpage(lastitem);
	 if ((lastitem + 1) < (dot + 10))
            fillscreen();
         for (;lastitem < hp->nfields;){
	    itemlist[++lastitem] = (fb_upd *) fb_malloc(sizeof(fb_upd));
	    itemlist[lastitem]->fp = NULL;
	    itemlist[lastitem]->fv = NULL;
            if (lastitem >= dot + 10){
               dot += 10;
               fb_move(4, 1); fb_clrtobot(); fb_infoline();
               }
            fb_cx_push_env("6D", CX_KEY_SELECT, NIL);
            st = eitem(lastitem, 1);
            fb_cx_pop_env();
            if (st == FB_END){
	       fb_free((char *) itemlist[lastitem--]);
               break;
	       }
            }
      }

/* 
 *  eitem - edit an entire item 
 */
 
   static eitem(it, new)
      int it, new;

      {
         int row, i, st;
         char *fb_trim(), temp[FB_MAXLINE];
	 fb_field *fp = NULL;

         fb_move(cdb_t_lines, 1); fb_clrtoeol();
         if ((row = (it % 10)) == 0)
            row = 10;
         row = row * 2 + 2;
	 if (itemlist[it] == NULL)
	    fb_xerror(FB_BAD_DICT, hp->idict, "bad itemlist");
         if (new == 1)
            fput(it, 0);

         /* 
	  *  get id 
	  */
	  
         for (;;){
	    if (!cdb_choosefield){
	       if (new)
		  fb_fmessage("Enter FieldName, -=End");
	       else
		  fb_fmessage("Enter FieldName, <CTL>-X=Skip");
	       st = fb_input(row, col[1], FB_TITLESIZE, 0, FB_ALPHA, 
		    temp, FB_ECHO, FB_OKEND, FB_CONFIRM);
	       }
	    else{
	       st = fb_choosefield(temp);
	       }
            if (st == FB_END && new == 1){
               fb_move(row, 1); 
               fb_clrtoeol();
	       if (cdb_choosefield)
	          headers();
               return(FB_END);
               }
            else if (st == FB_END)
               st = FB_DEFAULT;
            if (st == FB_DEFAULT || (st == FB_ABORT && new == 1)){
	       if ((it - 1) < hp->nfields)
	          fp = cdb_kp[it - 1];
	       else
	          fp = cdb_kp[0];
	       }
            else if (st == FB_ABORT)
               break;
            else if (st == FB_AOK){
               fb_trim(temp);
	       fp = fb_findfield(temp, hp);
	       }
	    if (fp == NULL){
	       fb_screrr("Cannot locate that field. Sorry.");
	       continue;
	       }
	    if (fp->type == FB_FORMULA || fp->type == FB_LINK){
	       fb_screrr("Not here silly. Use dbdbas to change formulas/links.");
	       continue;
	       }
	    for(i = 1; i <= lastitem; i++)
	       if (itemlist[i]->fp == fp)
	          break;
	    if (i <= lastitem){
	       fb_screrr("That field is already in this update dictionary!");
	       continue;
	       }
            itemlist[it]->fp = fp;
	    break;
            }
	 if (cdb_choosefield){
	    headers();
	    if (!new){
	       fillscreen();
	       }
	    else{				/* must be new */
	       lastitem--;
	       fillscreen();
	       lastitem++;
	       fput(it, 0);	/* puts number on screen */
	       }
	    }
         fput(it, 1);
	 fput(it, 2);		/* type is always constant */
	 fput(it, 3);		/* size is always constant */
	 
	 /* 
	  *  get updates 
	  */
	  
         for(;;){
	    if (FB_OFNUMERIC(itemlist[it]->fp->type))
	       fb_fmessage("Enter Update (numeric/formula), <CTL>-X=Skip");
	    else
	       fb_fmessage("Enter Update (alphanumeric), <CTL>-X=Skip");
            st = fb_input(row, col[4], 49, 0,FB_ALPHA,temp,FB_ECHO,FB_NOEND,FB_CONFIRM);
            if (st == FB_ABORT && new != 1)
               strcpy(temp, itemlist[it]->fv);
	    else if (st == FB_DEFAULT || st == FB_ABORT)
	       strcpy(temp, " ");
	    if (check(fb_trim(temp), itemlist[it]->fp) != FB_AOK)
	       continue;
	    if (itemlist[it]->fv != NULL)
	       fb_free(itemlist[it]->fv);
	    itemlist[it]->fv = (char *) fb_malloc(strlen(temp)+1);
	    strcpy(itemlist[it]->fv, temp);
            break;
            }
         fput(it, 4);
	 
	 /*
	  * finally return
	  */

         return(FB_AOK);
      }

/* 
 *  check - check the type against the buffer for validity: 
 *    allows formulas for numeric fields 
 */
 
   static check(p, k)
      fb_field *k;
      char *p;
      
      {
         if (!(FB_OFNUMERIC(k->type))){
	    if (!equal(cdb_T_DELETE, p) && strlen(p) > k->size)
	       p[k->size] = NULL;
            return(FB_AOK);
	    }
	 return(fb_getformula(k, p, NIL, 1, hp));
      }

/* 
 *  fput - fb_put an entire item to a constant screen location 
 */
 
   static fput(i, pos)
      int i, pos;
      
      {
         int row;
         
         if ((row = (i % 10)) == 0)
            row = 10;
         row = row * 2 + 2;
         fb_move(row, col[pos]);
         switch (pos){
            case 0: fb_printw("%3d)", i); break;
            case 1: 
	       if (itemlist[i]->fp != NULL)
	          fb_printw("%-10s", itemlist[i]->fp->id);
	       else
	          fb_printw("<empty>");
	       break;
	    case 2: 
	       if (itemlist[i]->fp != NULL)
	          fb_printw(FB_FCHAR, itemlist[i]->fp->type);
	       break;
	    case 3: 
	       if (itemlist[i]->fp != NULL)
                  fb_printw("%6d", itemlist[i]->fp->size);
	       break;
	    case 4: 
	       fb_clrtoeol();
	       if (itemlist[i]->fv != NULL)
	          fb_printw(FB_FSTRING, itemlist[i]->fv); 
	       break;
            }
      }

/* 
 *  newdictu - printout a new dictionary list of itemlist 
 */
 
   static newdictu(p)
      char *p;
      
      {
         int i;
	 FILE *fs, *fb_mustfopen();
	 
	 fs = fb_mustfopen(p, "w");
	 for (i = 1; i <= lastitem; i++)
	    if (itemlist[i]->fp != NULL){
	       fb_trim(itemlist[i]->fv);
	       fb_underscore(itemlist[i]->fp->id, 1); /* fb_put back underscores */
	       fb_underscore(itemlist[i]->fv, 1);     /* insert underscores */
	       fprintf(fs, "%s %s\n",
	             itemlist[i]->fp->id, itemlist[i]->fv);
	       }
	 fprintf(fs, "%%\n");
	 fclose(fs);
      }
      
/* 
 *  headers - print header at top of screen 
 */
 
   static headers()
      {
	 fb_scrstat("Define Updates");
         fb_move(2, 1); fb_clrtoeol();
	 fb_move(3, 1); fb_clrtoeol();
         fb_move(2,col[1]);   fb_printw("Field");
         fb_move(3,col[1]);   fb_printw("=====");
	 fb_move(2,col[2]-1); fb_printw("Type");
	 fb_move(3,col[2]-1); fb_printw("====");
	 fb_move(2,col[3]+2); fb_printw("Size");
	 fb_move(3,col[3]+2); fb_printw("====");
	 fb_move(2,col[4]);   fb_printw("Update String");
	 fb_move(3,col[4]);   fb_printw("=============");
      }
      
/* 
 *  usage message 
 */
 
   usage()
      {
         fb_xerror(FB_MESSAGE, "usage: dbdupd [-d dbase] [-i index]", NIL);
      }
